### ISSUE TITLE
MetalConfig quantization: batched generate silently corrupts all rows after the first — affine_qmm_t computes only batch element 0 of 3D inputs (verified, one-line fix)

### ISSUE THREAD

**Issue by fenilsonani** [huggingface/transformers#47331]
### System Info

```
- `transformers` version: 5.13.1
- Platform: macOS-26.5.2-arm64-arm-64bit
- Python version: 3.11.15
- Huggingface_hub version: 1.23.0
- Safetensors version: 0.8.0
- Accelerate version: 1.14.0
- Accelerate config: not found
- DeepSpeed version: not installed
- PyTorch version (accelerator?): 2.13.0 (MPS)
- kernels version: 0.15.2
- Device: Apple M4 Max, 36 GB, MPS backend
- Using distributed or parallel set-up in script?: No
```

### Who can help?

@MekkCyber @vasqu

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder
- [x] My own task or dataset (give details below)

### Reproduction

Models quantized with `MetalConfig` (bits=4 and bits=8 identical) generate correctly at
batch size 1, but at batch size > 1 only the first row is correct; every later row's
argmax collapses to token 0 (decoding as `!!!!...`). Deterministic — repeated runs are
bit-identical.

**Model-level repro** (greedy, same prompt duplicated — every row should match batch 1):

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer, MetalConfig

MODEL = "Qwen/Qwen3-0.6B"
tok = AutoTokenizer.from_pretrained(MODEL)

def generate(model, batch):
    ids = tok(["The capital of France is"] * batch, return_tensors="pt").to("mps")
    with torch.inference_mode():
        out = model.generate(**ids, max_new_tokens=20, do_sample=False)
    return [tok.decode(o[ids["input_ids"].shape[1]:], skip_special_tokens=True) for o in out]

q = AutoModelForCausalLM.from_pretrained(
    MODEL, quantization_config=MetalConfig(bits=8), dtype=torch.bfloat16
).eval().to("mps")
print(generate(q, 2))
# ['" Paris. The capital of Italy is Rome. ..."', '"!!!!!!!!!!!!!!!!!!!!"']
# row 0 == batch-1 output exactly; row 1 garbage. Unquantized bf16: all rows correct.
```

**Root cause (verified at kernel level, no model involved).** `MetalLinear.forward`
passes its input to `affine_qmm_t` as-is; during HF decode that input is 3D
`[batch, 1, hidden]`. Calling the kernel directly with a random quantized weight and
checking each output row against `x @ dequant(W).T`:

| input shape | result |
|---|---|
| `[B, K]` 2D, any B | all rows correct |
| `[1, S, K]` any S | all rows correct |
| `[B, 1, K]`, B > 1 | **batch element 0 correct, elements 1+ garbage** |
| `[2, 4, K]` | **rows 0–3 (batch elem 0) correct, rows 4–7 (elem 1) garbage** |

Identical pattern for bits=4/8 × bf16/fp16. The `[2, 4, K]` case pins it: the kernel
computes the **entire first batch element** and never touches the rest — it treats the
input as 2D `[shape[-2], K]`, ignoring the leading batch dimension. The garbage rows are
bit-identical across repeated calls. Probe script:

```python
import torch
from transformers.integrations.metal_quantization import (
    _affine_quantize_tensor, _affine_dequantize_tensor, _get_metal_kernel)

kernel = _get_metal_kernel()
torch.manual_seed(0)
K, N, GROUP, BITS = 2048, 1024, 64, 4
wp, sc, bi = _affine_quantize_tensor(torch.randn(N, K), GROUP, BITS)
w_deq = _affine_dequantize_tensor(wp, sc, bi, GROUP, BITS).to("mps")
wp, sc, bi = wp.to("mps"), sc.to("mps").bfloat16(), bi.to("mps").bfloat16()

for shape in [(2, K), (2, 1, K), (2, 4, K)]:
    x = torch.randn(*shape, device="mps", dtype=torch.bfloat16)
    y = kernel.affine_qmm_t(x, wp, sc, bi, GROUP, BITS)
    ref = (x.reshape(-1, K).float() @ w_deq.t()).reshape(*shape[:-1], N)
    err = ((y.float() - ref).abs().amax(-1) / ref.abs().amax(-1)).flatten()
    print(shape, [f"{e:.3f}" for e in err.tolist()])
# (2, 2048)       ['0.005', '0.006']            <- 2D fine
# (2, 1, 2048)    ['0.006', '9.874']            <- 3D: only batch elem 0 computed
# (2, 4, 2048)    ['0.005', ..., '9.9e+30', ...] <- rows 0-3 fine, 4-7 garbage
```

**Fix (verified).** Flattening 3D inputs to 2D around the kernel call makes batched
generate fully correct — with this patch all batch-4 rows match the batch-1 output
exactly:

```python
def forward(self, input):                      # MetalLinear.forward
    if input.dim() == 3:
        b, s, k = input.shape
        return _orig_forward(self, input.reshape(b * s, k)).reshape(b, s, -1)
    return _orig_forward(self, input)
```

Either `MetalLinear.forward` should flatten to 2D before calling `affine_qmm_t`
(one line, no kernel change needed — happy to open a PR), or the kernel should handle
the batch dimension of 3D inputs.

**Possibly relevant:** the kernel repo ships builds for torch 2.8 / 2.9 / 2.10 only
(`build/torch{28,29,210}-metal-aarch64-darwin`); under torch 2.13 the `kernels` loader
selects the `torch210` build. Reproduced on that combination.

**Impact:** any batched `generate()` with a Metal-quantized model on MPS silently
corrupts all sequences after the first — no error, no warning. Found while benchmarking
nvidia/canary-qwen-2.5b (speech LLM): int4 batch-32 transcription WER went from 2.1%
(batch 1) to 98% (batch 32). Batch 1 is unaffected.

### Expected behavior

Every row of a batched `generate()` should produce the same output as batch size 1 for
the same (duplicated, greedy-decoded) prompt — as batch element 0 already does, as the
2D-flatten patch produces, and as the unquantized bf16 model does for all rows.

**Comment by Rocketknight1** [comment]
cc @sunmarc for quants

**Comment by varneeth01** [comment]
Hi! I'd like to work on this issue if it's still available.
I'll investigate the Quark API changes, verify compatibility across supported versions, and prepare a fix with tests if appropriate.
If no one is already working on it, I'll open a draft PR once I have a working implementation.

**Comment by SunMarc** [comment]
Can you check if this is a torch version issue by testing with torch 2.10 ? We didn't update the kernels for newer version of torch as you saw. IF this is not the case, I think the best solution would be to fix the kernels so that the 3d case is correctly supported. Would you like to give it a try @fenilsonani ?

**Comment by fenilsonani** [comment]
Hey @SunMarc i already tested with pytorch 2.10 and other versions too and it reproduces exactly the same so it's not a version issue. and happy to try kernel fix.

**Comment by fenilsonani** [comment]
Hey @SunMarc i found the root cause and have the fix ready. the bug is in the host dispatch code in the kernels-community repo in quantized.mm the set_strides helper passes the row stride of scales/biases where the kernel expects the batch stride so every batch element after 0 reads scales shifted by z rows and for high rows it reads past the end of the tensor that's why row 0 is always correct and the rest is garbage. i verified this before touching any code by predicting the garbage exactly with row shifted scales and it matched the actual kernel output to 0.4%.

while testing i also found affine_qmv batched is broken in a different way the host passes ndim-1 batch dims but the kernel expects ndim-2 and reads M from x_shape[batch_ndim] which is out of bounds and the rows are covered by tid.x which the grid never set. and also the batched grid.z path is very slow even when correct because every batch element pays a full 32 row tile for M=1 so [32,1,K] on the lm_head shape was 60ms vs 2.2ms as a flattened 2d gemm so i added a dense batch collapse in the public ops that reshapes contiguous batched input into one 2d gemm and keeps the strided path as fallback and also a contiguity guard because the kernels assume the matrix dims are row contiguous and nothing enforced it.

tested on torch 2.10 and 2.13 with the native build for each all shapes from 1d to 5d including sliced and transposed batch dims bits 4/8 fp16/bf16 all pass and batch 1 output is bitwise identical to the shipped binaries so no regression. also ran batch 4 generate end to end with the fixed kernel and no transformers change and every row matches batch 1. opening a pr on huggingface/kernels-community with the fix and the batched tests will link it here.

**Comment by fenilsonani** [comment]
Hey @SunMarc the pr is up https://github.com/huggingface/kernels-community/pull/1038 but the vouch bot closed it right away since i am not on the vouched list can you add me or reopen it? everything from my last comment is in there plus the qmv dispatch fix and the batched tests and batch 1 stays bitwise identical to the shipped builds so no regression.

**Comment by francip** [comment]
Independently reproduced on the current stable stack. Could this issue be reopened? This is still reproducible with the **native torch 2.14 Metal kernel build**, not an older torch build selected as a fallback.

Environment:

| Component | Version |
|---|---|
| Hardware | Apple M4 Pro, 64 GiB unified memory |
| OS | macOS 26.6.2, arm64 |
| Python | CPython 3.12.13 |
| PyTorch | 2.14.0, MPS available |
| Transformers | 5.16.1 |
| kernels / kernels-data | 0.16.1 / 0.16.1 |
| huggingface-hub | 1.30.0 |
| tokenizers | 0.23.2 |
| safetensors | 0.8.0 |
| accelerate | 1.14.0 |
| NumPy | 2.5.2 |

Kernel: `kernels-community/mlx-quantization-metal-kernels`, revision `7b797756cb3970a16d1b700d8c2d7411b532e3e2`, build `torch214-metal-aarch64-darwin`. Native library SHA256: `67f4a136e5efa6edb2dfaddf207dee9710211de6ab16053c3780bd7c718ba838`. This is the artifact selected by the installed Transformers integration (`version=1`). Our separate model comparison also used MLX/MLX-Metal 0.32.2 and MLX-LM 0.31.3; none is imported by this reproducer.

Minimal reproduction: no model download, tokenizer, adapter, padding, or private data. The kernel package is downloaded on first use.

```sh
uv venv .venv --python 3.12
uv pip install --python .venv/bin/python torch==2.14.0 transformers==5.16.1 kernels==0.16.1 huggingface-hub==1.30.0
.venv/bin/python repro.py
```

```python
import torch
from transformers.integrations.metal_quantization import (
    MetalLinear, _affine_quantize_tensor,
)

assert torch.backends.mps.is_available()
torch.manual_seed(42)
K, N, GROUP, BITS = 512, 256, 64, 8
w, s, b = _affine_quantize_tensor(torch.randn(N, K), GROUP, BITS)
layer = MetalLinear(K, N, bits=BITS, group_size=GROUP).to("mps")
with torch.no_grad():
    layer.weight.copy_(w)
    layer.scales.copy_(s)
    layer.qbiases.copy_(b)
x = torch.randn(4, 3, K, device="mps", dtype=torch.bfloat16)
with torch.inference_mode():
    actual = layer(x)
    flat = layer(x.reshape(-1, K)).reshape(4, 3, N)
    rows = torch.stack([layer(row) for row in x])
torch.mps.synchronize()
print("3D finite by batch:", torch.isfinite(actual).all(-1).all(-1).tolist())
print("flattened finite:", bool(torch.isfinite(flat).all()))
print("3D vs individual max abs error by batch:",
      (actual.float() - rows.float()).abs().amax((1, 2)).tolist())
print("flattened vs individual max abs error:",
      (flat.float() - rows.float()).abs().max().item())
torch.testing.assert_close(flat, rows, rtol=0.02, atol=0.125)
# Fails without the integration-side flatten/restore fix.
torch.testing.assert_close(actual, rows, rtol=0.02, atol=0.125)
```

The flattened and individually evaluated batches match exactly in our run. The unmodified 3-D call fails the final assertion, with very large errors after batch element 0. Garbage values depend on the run/allocation; a finite-value check alone is not sufficient. In our full-model LFM2.5-230M runs, some of these rows also produced non-finite logits. Unquantized BF16 did not exhibit this failure.

We verified a local integration-side flatten/restore workaround on batched inference and retained a separate per-row isolation check. We saw the earlier kernel PR huggingface/kernels-community#1038 and integration PR #47970; this comment is additional current-version evidence, not a claim to have discovered the issue first.

**Comment by francip** [comment]
A tested two-line integration fix and regression tests are ready on [this branch](https://github.com/kortexa-ai/transformers/tree/kortexa-ai/fix-metal-batched-inputs) ([signed commit](https://github.com/kortexa-ai/transformers/commit/fd3b8c2dabb22917a729ed97136e1927b3d3c141)), based on `main` at `c93057d4835c`. The exact reproducer above fails on unmodified `main` and passes with zero batch-versus-individual error after flatten/restore.

`TRANSFORMERS_TEST_DEVICE=mps .venv/bin/python -m pytest tests/quantization/metal/test_metal.py -q`: 58 passed, 2 slow tests skipped, 60 subtests passed on the native torch 2.14 kernel. Coverage includes 1D–4D, transposed batches, 4/8 bits, FP16/BF16, and bias. All 17 `make fix-repo` checks pass in patch mode; a separate type check has the same five existing diagnostics on base and patched source.

This patch and its tests were prepared with AI assistance. I am not a vouched contributor. Credit to the original report, #47970, and huggingface/kernels-community#1038; this follows the same integration approach with current native-kernel evidence and real-MPS coverage. No PR is open because the contribution rules require prior approval. Would the issue author or a maintainer approve an upstream PR for this fix?

**Comment by vasqu** [comment]
gentle ping @SunMarc
